### PR TITLE
add: Performance Options & automatic optimization

### PR DESCRIPTION
**Pull request by morganholly**
add ability to run every N frames
add ability to change frame divisor based on forces applied (makes it mostly invisible)

only tested with small graph, but this resulted in going from maxing out my potato gpu and thermal throttling my cpu to having no visible hit

### CODE REVIEW

**Comment by morganholly** [comment]
the automatic option to run less often when the graph is not changing much ends up running at only once every 20 frames on the small graph after it stabilizes. on larger graphs the calculation might break down, but just changing the 8/total_force to a different number would change that and it would be a pretty simple addition to add another export parameter

File: addons/force_directed_graph/fdg_node.gd
Comment (ThePat02) [inline addons/force_directed_graph/fdg_node.gd:1]: Add and explain the return value to the docstring.

File: addons/force_directed_graph/fdg_spring.gd
Comment (ThePat02) [inline addons/force_directed_graph/fdg_spring.gd:1]: Same here! Just for keeping consistency!

File: addons/force_directed_graph/force_directed_graph.gd [hunk 1]
```diff
@@ -15,7 +15,12 @@ var springs: Array[FDGSpring]
 @export var is_active = true
 ## Whether the graph should be simulated in the editor.
 @export var simulate_in_editor = true
+## Run every N frames
+@export_range(0, 20) var frame_divide: int = 1
+## Use total applied force to calculate how often to update
+@export var auto_frame_divide: bool = true
```
Comment (ThePat02) [inline addons/force_directed_graph/force_directed_graph.gd:21]: Mention that this overwrites the value of `frame_divide`!

File: addons/force_directed_graph/force_directed_graph.gd [hunk 2]
```diff
@@ -15,7 +15,12 @@ var springs: Array[FDGSpring]
...
 
+var frame_counter = 1
```
Comment (ThePat02) [inline addons/force_directed_graph/force_directed_graph.gd]: I think a name more verbose like "frames_since_calculation" or similar would improve readability.

File: project.godot
Comment (ThePat02) [inline project.godot:1]: I will update to a newer Godot version in another commit. Doesn't make any difference tho.

**Review by ThePat02 — CHANGES_REQUESTED** [review]
Thank your for this PR! Besides some _minor_ naming and documentation changes, this is a great addition for the plugin. If you find some time to adjust your PR accordingly, I will gladly merge it into main!

Always happy to see people still use my old plugins :) (Especially **without** frying their GPU!!!)

**Comment by morganholly** [comment]
ah yea that makes sense, ill work on that today

**Comment by morganholly** [comment]
i added that additional documentation, i made it scale up forces when it runs less often (tho nonlinearly, to reduce jitter at very large divisions), and did some checking to look for editor dragging.

editor drag checking was the hardest part, since there was no notification or signal that would work as needed, but the checks i added do improve it a lot. it only checks that every 10 frames, which makes it not as smooth as it could be, but also reduces the performance hit. and when the drag gets going, it should update the total applied force enough to get it running most if not all frames.

**Comment by morganholly** [comment]
also i updated the range on the frame divide to go up to 500, which it is also now capped at in the code. it generally seems to stabilize between 100 and 500 when in a stable state.

**Review by ThePat02 — APPROVED** [review]

**Comment by ThePat02** [comment]
@morganholly Thank you for your contribution <3 Great work :)

**Comment by morganholly** [comment]
i will have to check it later but i just realized it might have a bug when the auto divide option is off, some of it at least still runs, i think

**Pull request merged**